### PR TITLE
Fixed the counterfactual generation for pytorch nn when using 'genetic' method

### PR DESCRIPTION
**Pull request by DanieleMorotti**
If you built a simple neural network with Pytorch for multiclass classification, you were not able to generate counterfactuals using the 'genetic' option, due to some shape mismatches. 
Changing the 2 files as I have done and selecting the argmax between the array of logits/probabilities for each sample, you can now generate the counterfactuals.

### CODE REVIEW

**Review by amit-sharma — APPROVED** [review]
thanks for this fix, @DanieleMorotti

**Comment by DanieleMorotti** [comment]
> thanks for this fix, @DanieleMorotti For some reason, the test suite is not working on this PR.

Which tests failed? Can I manually run them or should I use Github actions?
Thank you!

**Review by gaugup — CHANGES_REQUESTED** [review]
Could you add unit tests as well for this change?

**Comment by gaugup** [comment]
> To be honest it's my first pull request in a project like this, therefore I don't know what I should do. Sorry for that.

@DanieleMorotti, looks like tests are passing. If you add a unit test that will be great.

**Comment by amit-sharma** [comment]
agree with @gaugup you can add a test case in the `tests/test_dice_interface.py` and/or the `tests/test_explainer_base.py` file--this makes sure that in the future, this kind of error won't happen and will be caught by your test.

Then we are ready to merge. Thanks again for the contribution!

**Comment by DanieleMorotti** [comment]
I added a simple test that checks the shape of the predictions (the error that I encountered before the fix). Let me know if it's ok or if i need to change something. 
Thank you.

File: tests/test_dice_interface/test_dice_genetic.py [hunk 1]
```diff
@@ -177,7 +177,43 @@ def test_predict_custom(self, desired_class, sample_custom_query_2, mocker):
         mocker.patch('dice_ml.model_interfaces.base_model.BaseModel.get_output', return_value=[[0, 0.5, 0.5]])
         custom_preds = self.exp._predict_fn_custom(sample_custom_query_2, desired_class)
         assert custom_preds[0] == desired_class
-
+    
+    # Testing if the shapes of the predictions are correct for multiclass classification
+    @pytest.mark.parametrize(("desired_class", "method"), [(1, "genetic")])
+    def test_multiclass_nn(self, desired_class, method):
+        pytest.importorskip("torch")
+        pytest.importorskip("sklearn")
```
Comment (gaugup) [inline tests/test_dice_interface/test_dice_genetic.py:185]: Why are the above two lines needed? I think we install these dependencies before executing the tests.

File: tests/test_dice_interface/test_dice_genetic.py [hunk 2]
```diff
@@ -177,7 +177,43 @@ def test_predict_custom(self, desired_class, sample_custom_query_2, mocker):
...
+
+        backend = "PYT"
+        dataset = helpers.load_custom_testing_dataset_multiclass()
+
+        # Transform the categorical data to number to test the neural network
+        import sklearn
```
Comment (gaugup) [inline tests/test_dice_interface/test_dice_genetic.py:191]: Please move all imports to top of the file with other imports.

File: tests/test_dice_interface/test_dice_genetic.py [hunk 3]
```diff
@@ -177,7 +177,43 @@ def test_predict_custom(self, desired_class, sample_custom_query_2, mocker):
...
+
+        try:
+            # Test the function that returns the predictions
+            _, _, preds = exp.build_KD_tree(
+                df.copy(), desired_range=None, desired_class=desired_class, 
+                predicted_outcome_name=d.outcome_name + '_pred'
+            )
+            assert hasattr(preds, "shape"), "The returned object that contains the predictions doesn't have a 'shape' attribute."
+            assert preds.shape[0] == df.shape[0], "The number of predictions differ from the number of elements in the passed dataset."
+        except ValueError as ve:
```
Comment (gaugup) [inline tests/test_dice_interface/test_dice_genetic.py:213]: Why do we need this try and except in a test case? If the test is suppose to catch the actual regressions and we can just keep the code under try and let the executor of the test determine the cause of the error.

File: tests/test_dice_interface/test_dice_genetic.py [hunk 4]
```diff
@@ -177,7 +177,43 @@ def test_predict_custom(self, desired_class, sample_custom_query_2, mocker):
...
+            mess_error = ve.__repr__()
+            assert not 'length' in mess_error, "There is an error that is not related to the shape of the predictions."          
+
```
Comment (gaugup) [inline tests/test_dice_interface/test_dice_genetic.py:216]: Problem with the above code is if the test fails with 'length' in the exception message then your test fails for a different reason. Better recommendation is to remove the code in except.

**Review by gaugup — CHANGES_REQUESTED** [review]

File: tests/test_dice_interface/test_dice_genetic.py (hunk 1 above)
Comment (DanieleMorotti) [reply tests/test_dice_interface/test_dice_genetic.py:185]: I put the lines to make sure sklearn and pytorch were imported, but you're right. I'll remove them in the next commit.

File: tests/test_dice_interface/test_dice_genetic.py (hunk 2 above)
Comment (DanieleMorotti) [reply tests/test_dice_interface/test_dice_genetic.py:191]: I imported it within the function to avoid changing the global scope, but I will change this too, thanks.

File: tests/test_dice_interface/test_dice_genetic.py (hunk 4 above)
Comment (DanieleMorotti) [reply tests/test_dice_interface/test_dice_genetic.py:216]: The idea was to test only the functionality I changed, therefore I didn't want to capture other errors that may not be generated by problems in the shape. But you're right that we don't need the try and except, therefore I'll just leave the code that it's currently under the 'try'.

File: tests/test_dice_interface/test_dice_genetic.py (hunk 2 above)
Comment (amit-sharma) [reply tests/test_dice_interface/test_dice_genetic.py]: yeah, good to change. It will also resolve the linting error that github is showing:
> Run isort . -c
ERROR: /home/runner/work/DiCE/DiCE/tests/test_dice_interface/test_dice_genetic.py Imports are incorrectly sorted and/or formatted.

Comment (DanieleMorotti) [reply tests/test_dice_interface/test_dice_genetic.py]: After the changes in the commit I did yesterday, it seems that there was still an error in the linting. I run isort locally and it removes a whitespace, let's see if it was the problem.

**Review by gaugup — APPROVED** [review]

**Comment by gaugup** [comment]
@DanieleMorotti, approved your changes. Feel free to merge if the gated tests pass.

**Comment by amit-sharma** [comment]
flake8 is detecting some errors. can you fix the following errors from the build?
you can use `flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics` to reproduce locally. 

> ./dice_ml/utils/neuralnetworks.py:39:19: W292 no newline at end of file
./tests/test_dice_interface/test_dice_genetic.py:181:1: W293 blank line contains whitespace
./tests/test_dice_interface/test_dice_genetic.py:203:72: W291 trailing whitespace
./tests/test_dice_interface/test_dice_genetic.py:207:47: F541 f-string is missing placeholders
./tests/test_dice_interface/test_dice_genetic.py:207:128: E501 line too long (132 > 127 characters)
./tests/test_dice_interface/test_dice_genetic.py:208:1: W293 blank line contains whitespace
1     E501 line too long (132 > 127 characters)
1     F541 f-string is missing placeholders
1     W291 trailing whitespace
1     W292 no newline at end of file
2     W293 blank line contains whitespace

File: tests/test_dice_interface/test_dice_genetic.py (hunk 2 above)
Comment (amit-sharma) [reply tests/test_dice_interface/test_dice_genetic.py]: this is due to flake8. see my comment below.

**Comment by DanieleMorotti** [comment]
> flake8 is detecting some errors. can you fix the following errors from the build? you can use `flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics` to reproduce locally.
> 
> > ./dice_ml/utils/neuralnetworks.py:39:19: W292 no newline at end of file
> > ./tests/test_dice_interface/test_dice_genetic.py:181:1: W293 blank line contains whitespace
> > ./tests/test_dice_interface/test_dice_genetic.py:203:72: W291 trailing whitespace
> > ./tests/test_dice_interface/test_dice_genetic.py:207:47: F541 f-string is missing placeholders
> > ./tests/test_dice_interface/test_dice_genetic.py:207:128: E501 line too long (132 > 127 characters)
> > ./tests/test_dice_interface/test_dice_genetic.py:208:1: W293 blank line contains whitespace
> > 1     E501 line too long (132 > 127 characters)
> > 1     F541 f-string is missing placeholders
> > 1     W291 trailing whitespace
> > 1     W292 no newline at end of file
> > 2     W293 blank line contains whitespace

Thank you! I pushed the new changes, I hope this time I actually fixed the errors.

**Comment by amit-sharma** [comment]
Perfect, merging it now
